Pt 157558948 contract chain test suite

rvirding opened the pull request:
This is the first proper version of the contract tests. It has been rebased on master a number of times and has incorporated all the necessary changes until now and works with them.

velzevur commented on apps/aehttp/test/aehttp_contracts_SUITE.erl:
```diff
@@ -0,0 +1,1504 @@
+%% ============================================================
+
+%% null(Config)
+%%  Does nothing an always succeeds.
```
This is probably a typo (`an` -> `and`?)

velzevur commented on apps/aehttp/test/aehttp_contracts_SUITE.erl:
```diff
@@ -0,0 +1,1504 @@
+    ct:pal("Balances 0: ~p, ~p\n", [ABal0,BBal0]),
+
+    %% Add tokens to both accounts.
+    %% {ok,200,_} = post_spend_tx(APubkey, 500, 1),
```
Why is this commented out?

velzevur commented on apps/aehttp/test/aehttp_contracts_SUITE.erl:
```diff
@@ -0,0 +1,1504 @@
+    Host = external_address(),
+    http_request(Host, post, "contract/decode-data", Request).
+
+%% get_spend(Data) ->
```
Probably we don't need the commented out functions now?

lucafavatella commented on apps/aehttp/test/aehttp_contracts_SUITE.erl:
```diff
@@ -0,0 +1,1234 @@
+		 acc_d => #{pub_key => DPubkey,
+			    priv_key => DPrivkey,
+			    start_amt => StartAmt}},
+    [{accounts,Accounts},{node_name,NodeName}|Config];
```
It seems that this init does not guarantee that Alice/Bert/Carl/Diana have StartAmt tokens, as it puts one spend tx per account, having a sender the mining node, in the mempool though without checking inclusion in block. (This is neither good nor bad on its own - it depends what test case assumes.)

rvirding replied:
I have converted it to behave in exactly the same way is in the integration_SUITE.

lucafavatella commented on apps/aehttp/test/aehttp_contracts_SUITE.erl:
```diff
@@ -0,0 +1,1234 @@
+    %% Transfer money from Alice to Bert.
+    TxHash = spend_tokens(APubkey, APrivkey, BPubkey, 200, 5),
+    MineUntil = fun () -> tx_in_chain(TxHash) end,
+    aecore_suite_utils:mine_blocks_until(NodeName, MineUntil, 10),
```
I see this guarantees tx in chain.

lucafavatella commented on apps/aehttp/test/aehttp_contracts_SUITE.erl:
```diff
@@ -0,0 +1,1234 @@
+    %% Add tokens to both accounts.
+    {ok,200,_} = post_spend_tx(APubkey, 500, 1),
+    {ok,200,_} = post_spend_tx(BPubkey, 500, 1),
+    aecore_suite_utils:mine_blocks(NodeName, 2),
```
It seems that this does not guarantee that the two spend tx at the lines just above are in the chain. (This is neither good nor bad on its own - it depends what the rest of the test case assumes.)

rvirding replied:
This has been fixed with a mine_until.

lucafavatella commented on apps/aehttp/test/aehttp_contracts_SUITE.erl:
```diff
@@ -0,0 +1,1234 @@
+    ct:pal("Balances 2: ~p, ~p\n", [ABal2,BBal2]),
+
+    %% Check that the balances are correct, don't forget the fee.
+    ABal2 = ABal1 - 200 - 5,
```
This line is subject to intermittent test failure. This line assumes that `ABal1` reflects [the tx adding 500 tokens](https://github.com/aeternity/epoch/pull/1344/files#diff-c661915907572e4a612ffbdd1c98696eR183) though that is not guaranteed, as that tx may have been included only [while mining for the A->B transaction](https://github.com/aeternity/epoch/pull/1344/files#diff-c661915907572e4a612ffbdd1c98696eR195).

rvirding replied:
There are now relevant tests with the balances checking that they have the correct values.

lucafavatella commented on apps/aehttp/test/aehttp_contracts_SUITE.erl:
```diff
@@ -0,0 +1,1234 @@
+			 HexCode, <<"bid">>, <<"()">>, #{amount => 100000}),
+    ct:pal("Bid value ~p\n", [BidValue]),
+
+    %% These values look reasonable.
```
Could you please clarify whether the values of the balances are meant to be part of the passed/failed outcome of the test case? See also https://github.com/aeternity/epoch/wiki/Testing-general#debugging

rvirding replied:
This should have been the same type of test as in spending_1 but the gas complicated comparing the values. This will be fixed.

lucafavatella commented:
I took the liberty of labelling as WIP as I understand this PR may receive some changes.

rvirding commented:
@lucafavatella I have fixed your issues and cleaned up the code and started working with maps. I will commit this so it is ready and you can decide when it is best to merge it. I will continue with it until then.
